Fix multipart forms losing `boundary` values on edit

roytu opened the pull request:
#### Description

This is a fix for issue #4613 .  The problem is that in several places where multipart is used, the code assumes that the header will match `Content-Type: multipart/form-data` *exactly*, whereas the reality often has `Content-Type: multipart/form-data; boundary=---------foo`.

#### Checklist

 - [x] I have updated tests where applicable.
 - [x] I have added an entry to the CHANGELOG.

mhils commented on mitmproxy/contentviews/__init__.py:
```diff
@@ -101,7 +101,7 @@ def get_message_content_view(
         if ctype := message.headers.get("content-type"):
             if ct := http.parse_content_type(ctype):
-                content_type = f"{ct[0]}/{ct[1]}"
+                content_type = http.assemble_content_type(ct[0], ct[1], ct[2])
```
I believe we never need the content type attributes when selecting a contentview? Unless there's something I am missing, please revert the changes here.

mhils commented on mitmproxy/contentviews/multipart.py:
```diff
@@ -21,4 +21,7 @@ def __call__(self, data: bytes, content_type: Optional[str] = None, **metadata):
 
     def render_priority(self, data: bytes, *, content_type: Optional[str] = None, **metadata) -> float:
-        return float(content_type == "multipart/form-data")
+        if content_type and content_type.startswith("multipart/form-data"):
```
see above - this is only used to select the correct view, we don't need to handle the boundary information here.

mhils commented on mitmproxy/http.py:
```diff
@@ -944,7 +944,9 @@ def _get_multipart_form(self):
-        self.headers["content-type"] = "multipart/form-data"
+        if "content-type" not in self.headers:
+            # Don't overwrite header if it already exists or it will destroy the boundary value
+            self.headers["content-type"] = "multipart/form-data"
```
This... looks like it plainly didn't work before? Good catch.

I would suggest we change the logic here to 

1. Check if `self.headers["content_type"].startswith("multipart/form-data")`, and if that's not the case, add a content-type header with a random (?) boundary. The point here is that if someone assigns to `.multipart_form = ` in their addon, the header should be updated accordingly if the request was not multipart before.
2. *After* that, call `self.content = multipart.encode(self.headers, value)` (as we do right now)

mhils reviewed (changes requested):
Thanks! This looks like a good catch. I've added some comments below. In short, we can take contentviews out of the picture, and then change the multipart setter to actually work. 😃

roytu commented:
Thanks for the comments :)

Added commits to revert the `contentviews` stuff and the fix still seems to work, so I believe you're correct that the content type attributes aren't used there.

Also updated the `set_multipart_form` function to generate a boundary if the header doesn't exist (stole it from https://github.com/urllib3/urllib3/blob/2ef05adfcf001c85f1f95ae3fa078d16190851b0/src/urllib3/filepost.py#L20-L24 and added a few dashes for readability).

mhils reviewed (approved):
This looks fantastic now, thanks for the quick turnaround! 🍰 😃
